pkg/operator/reconciler.go

```diff
@@ -0,0 +1,246 @@
+package operator
+
+import (
+	"context"
+	"os"
+	"reflect"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	errors "k8s.io/apimachinery/pkg/api/errors"
+	k8serros "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	tunedv1 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/tuned/v1"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/manifests"
+	ntomf "github.com/openshift/cluster-node-tuning-operator/pkg/manifests"
+)
+
+// Reconciler reconciles a Tuned object and creates the tuned daemon set.
+type Reconciler struct {
+	Client    client.Client
+	Scheme    *runtime.Scheme
+	Namespace string
+}
+
+// SetupWithManager creates a new ClusterOperator Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// We want to initiate reconcile loop only on changes under the NTO cluster operator.
+	cvoPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object == nil {
+				klog.Error("create event has no runtime object to create")
+				return false
+			}
+
+			return e.Object.GetName() == tunedv1.TunedClusterOperatorResourceName
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !validateUpdateEvent(&e) {
+				return false
+			}
+
+			return e.ObjectNew.GetName() == tunedv1.TunedClusterOperatorResourceName &&
+				e.ObjectOld.GetName() == tunedv1.TunedClusterOperatorResourceName
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			if e.Object == nil {
+				klog.Error("delete event has no runtime object to delete")
+				return false
+			}
+
+			return e.Object.GetName() == tunedv1.TunedClusterOperatorResourceName
+		},
+	}
+
+	dsUpdatePredicate := predicate.Funcs{UpdateFunc: func(e event.UpdateEvent) bool {
+		if !validateUpdateEvent(&e) {
+			return false
+		}
+
+		dsOld := e.ObjectOld.(*appsv1.DaemonSet)
+		dsNew := e.ObjectNew.(*appsv1.DaemonSet)
+
+		return !equality.Semantic.DeepEqual(dsOld.Status, dsNew.Status) ||
+			!equality.Semantic.DeepEqual(dsOld.Spec, dsNew.Spec)
+	}}
+
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&configv1.ClusterOperator{}, builder.WithPredicates(cvoPredicates)).
+		Owns(&appsv1.DaemonSet{}, builder.WithPredicates(dsUpdatePredicate)).
+		Owns(&tunedv1.Tuned{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Complete(r)
```

Comment:
The predicates are useful when we want to filter the watching, e.g.:

I want to watch all Pods created/owned by this controller but I just want to re-trigger the reconciliation when they are updated if they have the label = n. It helps improve the performance since is possible create conditionals.